mat64/index_bound_checks.go

```diff
@@ -63,3 +65,47 @@ func (m *Vector) set(r int, v float64) {
 	}
 	m.mat.Data[r*m.mat.Inc] = v
 }
+
+// At returns the element at row r and column c.
+func (t *Symmetric) At(r, c int) float64 {
+	return t.at(r, c)
+}
+func (t *Symmetric) at(r, c int) float64 {
+	if r >= t.mat.N || r < 0 {
+		panic(ErrRowAccess)
+	}
+	if c >= t.mat.N || c < 0 {
+		panic(ErrColAccess)
+	}
+	if t.mat.Uplo == blas.Upper {
+		if r > c {
+			r, c = c, r
+		}
+		return t.mat.Data[r*t.mat.Stride+c]
+	}
+	if r < c {
+		r, c = c, r
+	}
+	return t.mat.Data[r*t.mat.Stride+c]
+}
+
+// Set sets the element at row r and column c. Set panics if the location is outside
+// the appropriate half of the matrix.
```

Comment:
As to your last point, if the function signature is SetSym, the user will be very aware that they're setting two things. Access will always work as expected.

Your first point is a very good one. It's not clear to me why blas allows both orientations of the matrix. My guess would be that there's some trickery one can do to save memory by treating half a Dense as symmetric. I'm not sure though -- it's not like you can fit two symmetric into one Dense. I do like the simplicity from having only Upper rather than needing to support both.